codersdk/audit.go

```diff
@@ -0,0 +1,53 @@
+package codersdk
+
+import (
+	"encoding/json"
+	"net/netip"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+type ResourceType string
+
+const (
+	ResourceTypeOrganization    ResourceType = "organization"
+	ResourceTypeTemplate        ResourceType = "template"
+	ResourceTypeTemplateVersion ResourceType = "template_version"
+	ResourceTypeUser            ResourceType = "user"
+	ResourceTypeWorkspace       ResourceType = "workspace"
+)
+
+type AuditAction string
+
+const (
+	AuditActionCreate AuditAction = "create"
+	AuditActionWrite  AuditAction = "write"
+	AuditActionDelete AuditAction = "delete"
+)
+
+type AuditDiff map[string]struct {
+	Old    any
+	New    any
+	Secret bool
+}
+
+type AuditLog struct {
+	ID               uuid.UUID       `json:"id"`
+	RequestID        uuid.UUID       `json:"request_id"`
+	Time             time.Time       `json:"time"`
+	OrganizationID   uuid.UUID       `json:"organization_id"`
+	IP               netip.Addr      `json:"ip"`
+	UserAgent        string          `json:"user_agent"`
+	ResourceType     ResourceType    `json:"resource_type"`
+	ResourceID       uuid.UUID       `json:"resource_id"`
+	ResourceTarget   string          `json:"resource_target"`
+	Action           AuditAction     `json:"action"`
+	Diff             AuditDiff       `json:"diff"`
+	StatusCode       int32           `json:"status_code"`
+	AdditionalFields json.RawMessage `json:"additional_fields"`
+	Description      string          `json:"description"`
+
+	User     *User           `json:"user"`
+	Resource json.RawMessage `json:"resource"`
```

Comment:
this field doesn't appear to be in the database table.  How is it persisted?

Reply:
I believe the idea in the RFC is that we would only persist `AdditionalFields` and not the entire object. This is the full resource object, populated at query time.

Reply:
The audit logs are historical, and so we need to persist everything.  If you populate at query time you get the state of the object _at query time_, rather than what you want which is the state of the object when the audit log dropped.